src/networks/competetive_net.py

```diff
@@ -238,4 +262,217 @@ def compute_loss_against(self, opponent, input):
+            self.optimize_bias,
+            mnist_28x28_conv=self.mnist_28x28_conv
+        )
+
+    def _get_labeled_mask(self, batch_size):
+        label_mask = to_pytorch_variable(torch.zeros(batch_size))
+        label_count = to_pytorch_variable(torch.tensor(batch_size * self.label_rate).int())
+        label_mask[range(label_count)] = 1.0
+        return label_mask
+
+    def _log_classification_distribution(self, ground_truth, label_mask, labels,
+                                         num_usable_labels, pred_labels):
+        frequencies = labels.bincount(minlength=self.num_classes).float()
+        frequencies[frequencies == 0.0] = 1e-10
+        frequencies = frequencies.data.cpu().numpy()
+        usable_labels = label_mask * labels.float()
+
+        label_data_per_class = usable_labels.long().bincount(
+            minlength=self.num_classes
+        )
+        label_data_per_class[0] = num_usable_labels - label_data_per_class[1:].sum()
+        predicted_frequencies = np.bincount(
+            pred_labels[np.where(pred_labels == ground_truth)],
+            minlength=self.num_classes
+        )
+        classification_rate_per_class = torch.from_numpy(
+            predicted_frequencies * 100.0 / frequencies
+        )
+
+        labels = to_pytorch_variable(torch.tensor([i for i in range(0, self.num_classes)]))
+        statistics = zip(labels, label_data_per_class, classification_rate_per_class)
+
+        stats = f"\nLabel, Number of Labeled Data points, Classification Rate for this label"
+        for (label, label_data, classification_rate) in statistics:
+            stats += f"\n{label}, {label_data}, {classification_rate}"
+        self._logger.info(stats)
+
+    def compute_loss_against(self, opponent, input, labels=None, alpha=None,
+                             beta=None, iter=None, log_class_distribution=False):
+        batch_size = input.size(0)
+        tensor = torch.Tensor(batch_size)
+        tensor.fill_(self.num_classes)
+        tensor = tensor.long()
+        fake_labels = to_pytorch_variable(tensor)
+
+        label_mask = self._get_labeled_mask(batch_size)
+
+        # Positive Label Smoothing
+        real = torch.Tensor(batch_size)
+        real.fill_(0.9)
+        real = to_pytorch_variable(real)
+
+        # Adding noise to prevent Discriminator from getting too strong
+        if iter is not None:
+            std = max(self.in_std_min, self.in_std - iter * self.in_std_decay_rate)
+        else:
+            std = self.in_std
+        input_perturbation = to_pytorch_variable(torch.empty(input.shape).normal_(mean=self.in_mean, std=std))
+        input = input + input_perturbation
+
+        if self.mnist_28x28_conv:
+            input = input.view(-1, 1, 28, 28)
+
+        network_output = self.classification_layer(self.net(input))
+        network_output = network_output.view(batch_size, -1)
+
+        # Real Supervised Loss
+        supervised_loss_function = CrossEntropyLoss(reduction='none')
```

Comment:
Why have you hardcoded the loss function? I saw that you have already created the configuration_container for them.
If you want, we can discuss it later.